cache/contenthash/filehash.go

```diff
@@ -41,6 +41,15 @@ func NewFileHash(path string, fi os.FileInfo) (hash.Hash, error) {
 }
 
 func NewFromStat(stat *fstypes.Stat) (hash.Hash, error) {
+	// Clear the irregular file bit if this is some kind of special
+	// file. Pre-Go 1.23 behavior would only add the irregular file
+	// bit to regular files with non-handled reparse points.
+	// Current versions of Go now apply them to directories too.
+	// archive/tar.FileInfoHeader does not handle the irregular bit.
+	if stat.Mode&uint32(os.ModeType&^os.ModeIrregular) != 0 {
```

Comment:
Isn't the negation wrong in the condition here?

Reply:
The intention is to use `os.ModeType` but to remove `os.ModeIrregular` because `os.ModeIrregular&os.ModeType` will always return non-zero. so this uses the `os.ModeType` constant but removes `os.ModeIrregular` from it.

It's a bit tricky because the original condition checks if `m&os.ModeType == 0` and sets `os.ModeIrregular` if it is. But, because we've now set `os.ModeIrregular`, the condition will always be true so the reverse doesn't work.